draft-ietf-ppm-dap.md

```diff
@@ -2587,6 +2587,62 @@ the task and all data pertaining to this task after `task_expiration`.
 Implementors SHOULD provide for some leeway so the Collector can collect the
 batch after some delay.
 
+### Distributed-systems and Synchronization Concerns
+
+Various parts of a DAP implementation will need to synchronize in order to
+ensure correctness during concurrent operation. This section describes the
+relevant concerns and makes suggestions as to potential implementation
+tradeoffs.
+
+* The upload interaction requires the Leader to ignore uploaded reports with a
+  duplicated ID, including concurrently-uploaded reports. This might be
+  implemented by synchronization or via an eventually-consistent process. If the
+  Leader wishes to alert the Client with a `reportRejected` error,
+  synchronization will be necessary to ensure exactly one concurrently-uploaded
+  report is accepted.
+
+* The Leader is responsible for generating aggregation jobs, and will generally
+  want to place each report in exactly one aggregation job. This may require
+  synchronization between different components of the system which are
+  generating aggregation jobs. Note that placing a report into more than one
+  aggregation job will result in a loss of throughput, rather than a loss of
+  correctness, privacy, or robustness, so it is acceptable for implementations
+  to use an eventually-consistent scheme which may rarely place a report into
+  multiple aggregation jobs.
```

Comment:
On its own, placing a report in multiple aggregation jobs would be bad for privacy, as it may allow a malicious helper to execute a replay attack by simply ignoring its own replay checks. (though it wouldn't be able to control what is replayed) Is the idea that the leader can use lax coordination to assemble aggregation jobs, and then catch the duplicates before it sends out the aggregation job initialization request, or before it finalizes the aggregation job and accumulates its output share?

Reply:
Yes -- grouping reports into aggregation jobs is separate from report-replay checking during aggregation [1]. The Leader could/should perform replay checking separately from aggregation job grouping, which would catch duplicates.

[1] Specifically, once a report ID is stored in the used-report storage used for replay checking, it can never be aggregated again. However, a report which is placed in an aggregation job, and then failed due to `report_too_early`, can be placed in another aggregation job later.

Reply:
> [1] Specifically, once a report ID is stored in the used-report storage used for replay checking, it can never be aggregated again. However, a report which is placed in an aggregation job, and then failed due to report_too_early, can be placed in another aggregation job later.

I've touched on this point with a few different folks at this point, WDYT about adding some explicit text indicating that used-report storage/aggregation replay checking should not be conflated with storage used to avoid placing the same report in multiple aggregation jobs? It might be a somewhat subtle point. (The only place this practically matters is due to `report_too_early`, but I don't think we want to remove those semantics in the name of simplicity.)

Reply:
Sure, that sounds good. Though, the two functions could be done by either one store with multiple states for each report, or by separate stores, depending on implementation choices. It seems like grouping reports into aggregation jobs should be easier to horizontally scale than the "used-report" part, without having to accept spurious duplication, by using  a distributed queue.

Reply:
Added an implementation note around this to the aggregation section; I think the phrasing doesn't contradict using the same storage system to store both bits of per-report state, it only notes that the storage used for replay protection shouldn't be used "directly" when generating new aggregation jobs.